trunk/src/lintool/lintool.1
.TH lintool "1" "March 2002" "gentoolkit 0.1.3" 
.SH NAME
lintool \- manual page for the lintool program, a program that checks the 
santity of ebuild scripts.
.SH SYNOPSIS
.B lintool
\fI<switches>\fR \fIebuilds...\fR
.SH DESCRIPTION
Lintool checks if a set of
.I ebuilds
conforms to the ebuild style guide. This is not
(yet) an exhaustive test, so your ebuild might be broken even if lintool 
thinks it's okay.
.PP
.SH OPTIONS 
.TP
\fB--no-summary\fI
Turn off total summary for all tests run on all ebuilds.
.TP
\fB--show-separate\fI
Show short summary of tests for each ebuild checked.
.TP
\fB--show-details\fI
Show full details of tests for each ebuild checked
.TP
\fB--tests=test1,test2,..\fI
Run only the tests specified to this option. Default is to run all
tests. One can turn off a particular test by pretending it with minus (e.g.
.I --tests=-TestUseFlags,-TestSpaces
). A list of tests can be obtained from
.I --list-tests
.TP
\fB--list-tests\fI
List available tests.
.SH AUTHORS
Karl Trygve Kalleberg <karltk@gentoo.org>, 2002
.SH "SEE ALSO"
ebuild(5) 
.TP
The \fI/usr/sbin/lintool\fR script. 



NEWS
News (new features/major bug fixes)

gentoolkit-0.6.5
----------------
  * eclean: fix eclean-dist with no git3-src dir (bug #922455)

gentoolkit-0.6.4
----------------
  * Raise portage requirement to >=3.0.57
  * Strip libc dependencies from eclean-pkg --changed-deps calculations


gentoolkit-0.6.3
----------------
  * Fix handling invalid binpkgs
  * Migrate to PEP517 meson-python build


gentoolkit-0.6.2
----------------
  * Add --clean-invalid option for eclean-pkg
  * Fix merge-driver-ekeyword compatibility with git >= 2.42
  * Fix eclean-pkg compatibility with FEATURES=pkgdir-index-trusted
  * Support the NO_COLOR environment variable


gentoolkit-0.6.1
----------------
  * Fix custom TMPL implementation for depends commit 3e35553df4673
  * fix unused variable detection
  * migrate setup.py to setuptools


gentoolkit-0.6.0
----------------
  * merge-driver-ekeyword Fix type annotations
  * equery changes removal
  * Convert from tab indent to spaces, format cleanup
  * Add github action to check formatting, add unit testing
  * Add flake testing to github action
  * ChangeLog cleanups
  * imlate: Remove 'experimental' code
  * Remove remnants of herds support
  * man page updates/fixes
  * euse: migrate egrep -> grep -E
  * Add python 3.10, 3.11 testing to tox.ini
  * implement --unique-use option for eclean-pkg
  * eshowkw: Fix bug 503366 row alignment issue
  * revdep-rebuild: set up a child logger with propagate = False
  * Add -f/--forced-masked to equery uses
  * metadata.py: Don't fail on unmatched package
  * equery: update linguas->l10n option
  * equery: Add -F TMPL option to depends module
  * revdep-rebuild.sh: use awk command instead of gawk
  * ekeyword: remove .ebuild file suffix requirement (bug 762331)
  * eclean: Complete migration to imported emaint code
  * eclean: add early return for empty DISTDIR
  * enalyze: Fix missing newline at end of saved rebuild files
  * equery: Turn off color when piping is set
  * equery meta: Fix missing description when not in metadata.xml
  * equery meta: Migrate to portage's MetadataXML class
  * equery check: Fix exception handling for Insufficient permissions
  * equery depends: Fix a traceback due to UnboundLocalError:


gentoolkit-0.5.1
----------------
  * Added support for Python 3.9
  * Removed support for Python < 3.6 and other clean ups
  * merge-driver-ekeyword: Added git merge driver for automatically resolving
    conflicts involving the KEYWORDS=... line
  * equery: Removed 'changes' subcommand
  * Enable ekeyword unit tests

gentoolkit-0.5.0
----------------
  * discontinued Support for Python 2
  * ekeyword and eshowkw now share logic for determining arch status
  * support for arches.desc to determine arch/~arch
  * updated fallback ~arch list with ~ia64 and ~m68k

gentoolkit-0.4.8
----------------
  * eclean packages: Add --changed-deps option to remove binpkgs
                     whose runtime depenedencies have changed
  * revdep-rebuild: Encode Popen arguments as utf8 (bug 709610)

gentoolkit-0.4.7
----------------
  * eshowkw: Add alpha to ~arch-only arches
  * Fixed deprecation warnings

gentoolkit-0.4.6
----------------
  * glsa-check moved to portage repository (bug 463952)
  * equery: fix incorrect status for forced flags (bug 634824)
  * revdep-rebuild: Fix pass_through_options to handle args containing
    spaces

gentoolkit-0.4.5
----------------
  * Include BDEPEND in dep checks

gentoolkit-0.4.4
----------------
  * eshowkw: Add riscv to ~arch-only arches

gentoolkit-0.4.3
----------------
  * epkginfo: Add 'reverse' display option for equery
  * Speed up "equery depends" by a great deal
  * Fix broken test suite and add new tests
  * Add tox config

gentoolkit-0.4.2
----------------
  * Improve keyword grouping and sorting in eshowkw
  * Bug fixes in eshowkw

gentoolkit-0.4.1
----------------
  * ekeyword: Support sorting keywords without additional changes
  * Bug fixes in revdep_rebuild, equery, eclean

gentoolkit-0.4.0
----------------
  * Merge ekeyword, ebump, imlate tools from gentoolkit-dev
  * Deprecate remaining gentoolkit-dev tools


gentoolkit-0.3.2
----------------
  * revdep-rebuild masks fixed
  * more unicode fixes
  * More deprecation warning fixes
  * glsa bug fix 400763 Apply listing only affected versions patch
  * eclean fixes and updates
  * eshowkw fixes and updates


gentoolkit-0.3.1
----------------
  * Make the python version of revdep-rebuild the default
  * more unicode fixes


gentoolkit-0.3.0
----------------

epkginfo:
  * is now a link to equery meta and has all the features equery meta has

equery:
  * --help menus cleaned up, using notations (b)elongs instead of belongs(b).
  * man page has been rewritten.
  * --quiet effects more modules.
  * 2 new modules:
    * changes - Gentoo ChangeLog viewer, try:
      `equery changes portage` to see entry for portage version that emerge
        wants to install;
      `equery changes portage --from=2.2_rc20 --to=2.2_rc30` to see all entries
        between the specified versions.
    * meta - Displays information available in metadata.xml and keyword info.
      Try `equery meta boost` to list herd, maintainers, keywords, and more.
  * Modules which are meant to run on multiple packages (check, list, size) now
    allow category and package name globbing, (so no more need for --exact-name
    or --category).

      # Exact name matching by default:
      $ equery l zilla
       * Searching for zilla ...

      # Use globs to fuzzy match
      $ equery l *zilla*
       * Searching for *zilla* ...
      [IP-] [  ] www-client/mozilla-firefox-3.5.4:0

      # Use globs to 'category filter'
      $ equery l www-client/*
       * Searching for * in www-client ...
      [I--] [XX] www-client/chromium-4.0.223.5:0
      [IP-] [  ] www-client/epiphany-2.26.3-r2:0
      [IP-] [  ] www-client/links-2.2:2
      [IP-] [  ] www-client/mozilla-firefox-3.5.4:0


deprecated/lintool/lintool.py
#! /usr/bin/python
#
# Copyright 2002 Gentoo Technologies, Inc
# Distributed under the terms of the GNU General Public License v2.0
# Author Karl Trygve Kalleberg <karltk@gentoo.org>
#
# About:
# lintool aims to check the stylistic and syntactical correctness for
# ebuilds, changelogs and digest files for the Gentoo packaging system.
#
# TODO
#
# - Make HTMLFormatter
#

VERSION="0.2.4"

import sys
import getopt

from lintool import ebuild, changelog, digest

class TextFormatter:
  def section(self, s):
    print "\n" + "-"*79
    print " " + s + "\n"
  def bullet(self, s):
    print "* " + s
  def sub(self, s):
    print "- " + s
  def subwarn(self, s):
    print "- (W) " + s
  def suberr(self, s):
    print "- (E) " + s
  def subsub(self, s):
    print " |" + s
  def subsubwarn(self, s):
    print " (W) |" + s
  def subsuberr(self, s):
    print " (E) |" + s
  def line(self,s):
    print s
  def div(self, left, right):
    l = len(left)
    r = len(right)
    return left + " " * (78-l-r) + right

class MunchieFormatter:
  def section(self, s):
    print "[lintool] " + "-" * (78 - len("[lintool] "))
    print "[lintool] " + s + "\n"
  def bullet(self, s):
    print "[lintool] * " + s
  def sub(self, s):
    print "[lintool] - " + s
  def subwarn(self, s):
    print "[lintool] - (W) " + s
  def suberr(self, s):
    print "[lintool] - (E) " + s
  def subsub(self, s):
    print "[lintool] |" + s
  def subsubwarn(self, s):
    print "[lintool]  (W) |" + s
  def subsuberr(self, s):
    print "[lintool] (E) |" + s
  def line(self,s):
    print "[lintool] " + s
  def div(self, left, right):
    l = len("[lintool] " + left)
    r = len(right)
    return left + " " * (78-l-r) + right

formatters = { "text" : TextFormatter(), "munchie" : MunchieFormatter()  }

def extractFilename(path):
  return path

def runTests(tests,results,ins):
  for j in tests:
    j.reset()

  ln = 1
  for i in ins.readlines():
    for j in tests:
      j.checkLine(i, ln)
    ln += 1

  hasWarning = 0
  hasError = 0
  for j in xrange(len(tests)):
    if tests[j].hasErrors():
      results[j][0] += 1
      hasError = 1
    if tests[j].hasWarnings():
      results[j][1] += 1
      hasWarning = 1
  return (hasError, hasWarning)
        
def showStatus(options,tests,formatter,file):
  if options['showDetails'] or options['showSeparate']:
    formatter.section("Status for " + file)
    for j in tests:
      if options['showSeparate'] or options['showDetails']:
        l = len(j.getDesc())
        formatter.bullet(formatter.div(j.getDesc(), ": " + j.getStatus()))
        if options['showDetails']:
          j.report()
  elif options['showShort']:
    allOK = 1
    for j in tests:
      if j.hasErrors():
        allOK = 0
        break
    if allOK:
      formatter.div(file, ": OK")
    else:
      formatter.div(file, ": Not OK")
    # else fall through the bottom    
    
def usage(opts):
  print sys.argv[0], "[options] ebuild [ebuild ebuild ... ]"
  print

  if opts:
    print "Where [options] include:"
    for (short,long_,desc) in opts:
      short_ = ''
      for s in short:
        short_ = short_ + '-' + s + ','
      long_ = '--' + long_
      opt = short_ + long_
      opt = opt.rjust(18)
      print opt + '  ' + desc
    print

def parse_opts(argv):
  options = { 'showSeparate': 0,
              'showTotal': 1,
              'showDetails': 1,
              'showShort': 1,
              'listTests': 0,
              'desiredTests': 0,
              'testMode' : "ebuild",
              'licenseDirs' : [ "/usr/portage/licenses" ],
              'formatter' : 'text'
            }

  opts = (('', 'show-separate',
           'Show short summary of tests for each ebuild checked'),

          ('v', 'version',
           'Show program version'),

          ('', 'no-summary',
           'Do not show total summary'),

          ('', 'no-details',
           'Do not show full details of tests for each ebuild checked'),

          ('', 'ebuild',
           'Files to check are ebuilds'),
          
          ('', 'changelog',
           'Files to check are changelogs'),

          ('', 'digest',
           'Files to check are digests'),

          ('', 'tests=',
           'Comma-separated list of tests to run'),

          ('', 'list-tests',
           'List available tests'),

          ('', 'from-file=<file>',
           'Read ebuilds from <file>'),

          ('', 'formatter=<formatter>',
           "Use 'text' (default) or 'munchie' formatter"),

          ('', 'aux-license-dir=<dir>',
           'Add <dir> to directories to search for licenses'),
            
          ('?h', 'help',
           'Show this help'),
         )

  short_options = ''
  long_options = []
  for (short,long_,desc) in opts:
    short_options = short_options + short
    if '=' in long_:
      long_ = long_.split('=', 1)[0] + '='
    long_options.append(long_)

  try:
    (option_list,args) = getopt.getopt(sys.argv[1:], short_options, long_options)
  except getopt.GetoptError, details:
    print 'Error parsing command line:',str(details)
    sys.exit(1)

  for (option,value) in option_list:
    if option in [ '--no-details' ]:
      options['showShort'] = 1
      options['showDetails'] = 0
    elif option in [ '--show-separate' ]:
      options['showShort'] = 0
      options['showSeparate'] = 1
    elif option in [ '--no-summary']:
      options['showTotal'] = 0
    elif option in [ '--from-file' ]:
      lines = open(value, 'r').readlines()
      lines = [o.strip() for o in lines]
      args = lines + args
    elif option in [ '--tests' ]:
      options['desiredTests'] = value.split(",")
    elif option in [ '--formatter' ]:
      options['formatter'] = value
    elif option in [ '--list-tests' ]:
      options['listTests'] = 1
    elif option in [ '--ebuild' ]:
      options['testMode'] = 'ebuild'
    elif option in [ '--changelog' ]:
      options['testMode'] = 'changelog'
    elif option in [ '--digest' ]:
      options['testMode'] = 'digest'
    elif option in [ '--aux-license-dir' ]:
      options['licenseDirs'].append(value)
    elif option in [ '-v', '--version' ]:
      print "Lintool " + VERSION
      sys.exit(0)
    elif option in [ '-h', '-?', '--help' ]:
      usage(opts)
      sys.exit(0)
    else:
      # shouldn't ever happen. better to be safe
      print "Unknown option - '%s'!" % (option)
      sys.exit(1)

  return (options,args)

def main():
    (options,args) = parse_opts(sys.argv[1:])

    formatter = formatters[options['formatter']]

    # Get test suite for given mode
    if options['testMode'] == "ebuild":
      available_tests = ebuild.getTests(formatter, options)
    elif options['testMode'] == "changelog":
      available_tests = changelog.getTests(formatter, options)
    elif options['testMode'] == "digest":
      available_tests = digest.getTests(formatter, options)

    # List available tests, if that was the users request
    if options['listTests']:
      maxlen = 0
      for i in available_tests:
        maxlen = max(len(i.__class__.__name__), maxlen)
      for i in available_tests:
        n = i.__class__.__name__
        print n + " " * (maxlen - len(n)) + " - " + i.getDesc()

    # Quit with short usage string, if no params given
    if len(args) == 0:
      usage(None)
      sys.exit(1)

    # Create final list of tests to run
    tests = []
    notTests = []
    if options['desiredTests']:
      for i in options['desiredTests']:
        for j in available_tests:
          if len(i) and i[0] == "-":
            notTests.append(i[1:])
          if j.__class__.__name__ == i:
            tests.append(j)
    else:
      tests = available_tests

    if len(notTests):
      for i in available_tests:
        if i.__class__.__name__ not in notTests:
          tests.append(i)
      
    results = [[0, 0] for x in range(len(tests))]

    # Set up for test run
    numFiles = 0
    totalErrors = 0
    totalWarnings = 0

    # Iterate through all files given as arguments, testing each file
    # against the final list of tests
    for i in args:
      fn = extractFilename(i)
      ins = open(i, "r")
      numFiles += 1
      (hasError, hasWarning) = runTests(tests,results,ins)
      totalErrors += hasError
      totalWarnings += hasWarning
      showStatus(options,tests,formatter,fn)

    # Show totals, if options allow it
    if options['showTotal']:
      formatter.section(formatter.div("Summary for all " + str(numFiles) + " " + options['testMode'] + "(s) checked", "#errors/warns"))
      for i in xrange(len(tests)):
        l = len(tests[i].getDesc())
        formatter.line(formatter.div(tests[i].getDesc(), ": %3d / %3d" % (results[i][0], results[i][1])))
      formatter.line(formatter.div("Total number of ebuilds with errors",  \
                                                   "%3d (%3d%%)" % (totalErrors, totalErrors*100/numFiles)))
      formatter.line(formatter.div("Total number of ebuilds with warnings", \
                                                  "%3d (%3d%%)" % (totalWarnings, totalWarnings*100/numFiles)))
    if totalErrors:
        sys.exit(1)

if __name__ == "__main__":
  main()

